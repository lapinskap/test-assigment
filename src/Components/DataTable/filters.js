import PropTypes from 'prop-types';
import React from 'react';
import __ from '../../utils/Translations';
import FilterDateRange from './FilterDateRange';
import Autocomplete from '../FormElements/Autocomplete/Autocomplete';

export const defaultFilterMethod = ({ id, value, pivotId }, row) => {
  if (!value) {
    return true;
  }
  const rowId = pivotId || id;
  const rowValue = row[rowId];
  if (Array.isArray(rowValue)) {
    return rowValue.map((el) => String(el).toLowerCase()).includes(String(value).toLowerCase());
  }
  return String(rowValue) && String(rowValue).toLowerCase().includes(String(value).toLowerCase());
};

export const SelectFilter = (options, submitForm = true) => ({ filter, onChange, column }) => {
  const optionsArray = options
    .map(({ label, value }) => <option key={value} value={value}>{__(label)}</option>);
  optionsArray.unshift(<option value="" key="empty-option">---</option>);
  return (
    <select
      data-t1="gridFilter"
      data-t2={column.id}
      className="form-control"
      value={filter ? filter.value : 0}
      onChange={(e) => {
        onChange(e.target.value);
        if (submitForm && e.target.form) {
          const submitButton = e.target.form.querySelector('button[type="submit"]');
          setTimeout(() => {
            submitButton.click();
          }, 10);
        }
      }}
    >
      {optionsArray}
    </select>
  );
};

export const AutocompleteSelectFilter = (options, submitForm = true) => ({ filter, onChange, column }) => (
  <Autocomplete
    id={column.id}
    options={options}
    classNamePrefix="gridFilterAutocomplete"
    value={filter ? filter.value : 0}
    onChange={(selected) => {
      onChange(selected?.value || '');
      if (submitForm) {
        const form = document.getElementById(column.id).closest('form');
        if (form) {
          const submitButton = form.querySelector('button[type="submit"]');
          setTimeout(() => {
            submitButton.click();
          }, 10);
        }
      }
    }}
  />
);

export const DefaultFilter = ({ filter, onChange, column }) => (
  <input
    data-t1="gridFilter"
    data-t2={column.id}
    type="text"
    style={{
      width: '100%',
    }}
    placeholder={column.Placeholder}
    value={filter ? filter.value : ''}
    onChange={(event) => onChange(event.target.value)}
  />
);

DefaultFilter.propTypes = {
  filter: PropTypes.shape({
    // eslint-disable-next-line react/forbid-prop-types
    value: PropTypes.any,
  }),
  onChange: PropTypes.func.isRequired,
  column: PropTypes.shape({
    id: PropTypes.string,
    Placeholder: PropTypes.string,
  }).isRequired,
};

DefaultFilter.defaultProps = {
  filter: null,
};

export const IntIdFilter = ({ filter, onChange, column }) => (
  <input
    data-t1="gridFilter"
    data-t2={column.id}
    type="text"
    style={{
      width: '100%',
    }}
    placeholder={column.Placeholder}
    value={filter ? filter.value : ''}
    onChange={(event) => onChange(event.target.value.replace(/\D/g, ''))}
  />
);

IntIdFilter.propTypes = {
  filter: PropTypes.shape({
    // eslint-disable-next-line react/forbid-prop-types
    value: PropTypes.any,
  }),
  onChange: PropTypes.func.isRequired,
  column: PropTypes.shape({
    id: PropTypes.string,
    Placeholder: PropTypes.string,
  }).isRequired,
};

IntIdFilter.defaultProps = {
  filter: null,
};

export const getArraySelectFilterMethod = (forceNumber = false) => ({ id, value }, row) => {
  if (!value) {
    return true;
  }
  return Array.isArray(row[id]) ? row[id].includes(forceNumber ? +value : value) : false;
};

export const booleanOptions = [{
  value: false,
  label: 'Nie',
}, {
  value: true,
  label: 'Tak',
}];

export const DateFilter = (hasTimeSelect = false) => ({ filter, onChange, column }) => (
  <FilterDateRange
    id={column.id}
    showTimeSelect={hasTimeSelect}
    value={filter ? filter.value : { from: '', to: '' }}
    onChange={onChange}
  />
);

export const dateFilterMethod = ({ id, value }, row) => {
  if (!value || (!value.to && !value.from)) {
    return true;
  }
  if (!row[id]) {
    return false;
  }

  const rowValue = new Date(row[id]);

  if (value.from && (new Date(value.from) > rowValue)) {
    return false;
  }

  if (value.to && (new Date(value.to) < rowValue)) {
    return false;
  }

  return true;
};

export const objectStatusOptions = [
  {
    value: 'active', label: 'Aktywny',
  },
  {
    value: 'inactive', label: 'Nieaktywny',
  },
  {
    value: 'new', label: 'Nowy',
  },
];

export const activeBooleanOptions = [{
  value: true,
  label: 'Aktywna',
}, {
  value: false,
  label: 'Nieaktywna',
}];
