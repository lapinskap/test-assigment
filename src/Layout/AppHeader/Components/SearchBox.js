import React from 'react';
import cx from 'classnames';

class SearchBox extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      activeSearch: false,
    };
  }

  render() {
    const { activeSearch } = this.state;
    return (
      <>
        <div className={cx('search-wrapper', {
          active: activeSearch,
        })}
        >
          <div className="input-holder">
            <input type="text" className="search-input" placeholder="Type to search" />
            <button
              type="button"
              onClick={() => this.setState({ activeSearch: !activeSearch })}
              className="search-icon"
            >
              <span />
            </button>
          </div>
          <button
            type="button"
            onClick={() => this.setState({ activeSearch: !activeSearch })}
            className="close"
          >
            <span />
          </button>
        </div>
      </>
    );
  }
}

export default SearchBox;
