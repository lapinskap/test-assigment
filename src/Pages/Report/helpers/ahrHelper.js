export const getAhrUrl = (url, isAhr) => {
  let result = isAhr ? '/ahr' : '';
  result += url;

  return result;
};

export const asd = () => {};
