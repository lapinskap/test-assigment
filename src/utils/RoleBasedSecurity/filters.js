import React from 'react';
import { godPermission } from './permissions';
import __ from '../Translations';
import ForbiddenPage from './SecurityComponents/ForbiddenPage';

export const filterNavItemsByAlc = (menu, userInfo) => {
  const permissions = userInfo.getPermissions();
  return menu.map((item) => {
    if (item.aclKey && !permissions.includes(item.aclKey) && !permissions.includes(godPermission)) {
      return null;
    }
    const updatedItem = { ...item };
    if (item.content) {
      updatedItem.content = filterNavItemsByAlc(item.content, userInfo);
      if (updatedItem.content.length === 0) {
        return null;
      }
    }
    return updatedItem;
  }).filter(Boolean);
};

export const filterTabsByAcl = (tabsConfig, userInfo) => {
  const permissions = userInfo.getPermissions();
  return tabsConfig.map((item) => {
    if (item.aclKey && !permissions.includes(item.aclKey) && !permissions.includes(godPermission)) {
      return {
        ...item,
        component: <ForbiddenPage permission={item.aclKey} />,
        disabled: true,
        tabClassName: 'not-allowed',
        tabTitle: __('Nie masz uprawnieÅ„ ({0})', [item.aclKey]),
      };
    }
    return item;
  }).filter(Boolean);
};

export const hasAccessTo = (userInfo, aclKey) => {
  const permissions = userInfo.getPermissions();
  return Boolean(permissions.includes(godPermission) || permissions.includes(aclKey));
};
