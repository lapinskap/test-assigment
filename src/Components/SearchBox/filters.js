import { AhrMainNav, MainNav } from '../../Layout/AppNav/NavItems';
import { filterNavItemsByAlc } from '../../utils/RoleBasedSecurity/filters';
import { mockData as mockCompanyData } from '../../Pages/Company/CompanyList';
import { mockData as mockEmployeeData } from '../../Pages/Company/EmployeeManagement/CompanyEmployees/List/list';
import __ from '../../utils/Translations';
import {
  CATALOG_MANAGEMENT_SERVICE, COMPANY_MANAGEMENT_SERVICE, restApiRequest, TOURISM_SERVICE,
} from '../../utils/Api';
import { productsMockData } from '../../Pages/ProductCatalog/List/table';
import { tourismObjectsMockData } from '../../Pages/Tourism/listOfObjects/table';

export const getResultsFromChildren = (items, value, resultHandler = []) => {
  let result = resultHandler;
  items.forEach(({ to, label, content }) => {
    if (to && __(label).toLowerCase().includes(value.toLowerCase())) {
      result.push({ to, label });
    }
    if (content && content.length) {
      result = getResultsFromChildren(content, value, resultHandler);
    }
  });
  return result;
};

export const findMenuItems = (value, ctx) => {
  if (!value || value.length < 2) {
    return [];
  }
  const items = filterNavItemsByAlc(MainNav, ctx.userInfo);
  return getResultsFromChildren(items, value).slice(0, 5).map(({ label, ...rest }) => ({ ...rest, label: __(label) }));
};

export const findEmployees = (value) => new Promise((resolve) => {
  setTimeout(() => {
    const result = mockEmployeeData.filter((item) => {
      const lowerCaseValue = value.toLowerCase();
      const name = `${item.firstName} ${item.lastName}`;
      return name.toLowerCase().includes(lowerCaseValue)
          || item.email.toLowerCase().includes(lowerCaseValue);
    }).map(({
      firstName, lastName, email, id,
    }) => ({
      label: `${firstName} ${lastName} (${email})`,
      to: `/ahr/employees/${id}`,
    }));
    resolve(result);
  }, 500);
});
export const findCompanies = (value) => restApiRequest(
  COMPANY_MANAGEMENT_SERVICE,
  '/companies',
  'GET',
  {
    params: {
      fullName: value,
      itemsPerPage: 10,
    },
  },
  mockCompanyData.filter((item) => item.fullName.toLowerCase().includes(value.toLowerCase())),
).then((res) => res.map(({ fullName, id }) => ({
  label: fullName,
  to: `/company/edit/${id}`,
})));
export const findProducts = (value) => restApiRequest(
  CATALOG_MANAGEMENT_SERVICE,
  '/products',
  'GET',
  {
    params: {
      name: value,
      itemsPerPage: 10,
    },
  },
  productsMockData.filter((item) => item.name.toLowerCase().includes(value.toLowerCase())),
).then((res) => res.map(({ name, id, mbProductType }) => ({
  label: name,
  to: `/product-catalog/products/${id}/${mbProductType}`,
})));
export const findTourismObjects = (value) => restApiRequest(
  TOURISM_SERVICE,
  '/tourism-objects',
  'GET',
  {
    params: {
      name: value,
      itemsPerPage: 10,
    },
  },
  tourismObjectsMockData.filter((item) => item.name.toLowerCase().includes(value.toLowerCase())),
).then((res) => res.map(({ name, id }) => ({
  label: name,
  to: `/tourism/objects/edit/${id}#basic_info`,
})));
