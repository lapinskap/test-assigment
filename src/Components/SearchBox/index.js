import React, {
  useState, useEffect, useContext, useRef, Fragment,
} from 'react';
import {
  Dropdown, DropdownItem, DropdownMenu, DropdownToggle,
} from 'reactstrap';
import { Link } from 'react-router-dom';
import { Loader as LoaderAnim } from 'react-loaders';
import RbsContext from '../../utils/RoleBasedSecurity/RbsContext';
import {
  findMenuItems, findCompanies, findEmployees, findProducts, findTourismObjects,
} from './filters';
import useHasPermission from '../../utils/hooks/security/useHasPermission';
import { companyCompanyPermissionRead } from '../../utils/RoleBasedSecurity/permissions';

export async function findData(inputText, setResult, filter) {
  if (inputText && inputText.length > 1) {
    try {
      const result = await filter(inputText);
      setResult(result);
    } catch (e) {
      console.error(e);
    }
  }
}

export default function SearchBox() {
  const [inputText, setInputText] = useState('');
  const [menuItems, setMenuItems] = useState([]);
  const [companies, setCompanies] = useState([]);
  const [products, setProducts] = useState([]);
  const [tourismObjects, setTourismObjects] = useState([]);
  const [employees, setEmployees] = useState([]);
  const [showSuggestions, setShowSuggestions] = useState(false);
  const [isSearching, setIsSearching] = useState(false);
  const rbsContext = useContext(RbsContext);
  const dropdown = useRef(null);
  const hasCompanyReadPermission = useHasPermission(companyCompanyPermissionRead);
  useEffect(() => {
    const handler = setTimeout(() => {
      if (inputText) {
        setIsSearching(true);
        setShowSuggestions(true);
        const newMenuItems = findMenuItems(inputText, rbsContext);
        setMenuItems(newMenuItems);
        const requests = [];
         if (hasCompanyReadPermission) {
          requests.push(findData(inputText, setCompanies, findCompanies));
          requests.push(findData(inputText, setProducts, findProducts));
          requests.push(findData(inputText, setTourismObjects, findTourismObjects));
        }
        Promise.all(requests).then(() => setIsSearching(false));
      } else {
        setShowSuggestions(false);
      }
    }, 500);
    return () => {
      clearTimeout(handler);
    };
  }, [inputText, rbsContext, hasCompanyReadPermission]);

  const toggle = (e) => {
    setShowSuggestions(dropdown.current.contains(e.target));
  };
  const foundSuggestions = Boolean(menuItems.length || companies.length || employees.length || products.length || tourismObjects.length);
  const suggestions = [
    { scope: 'Firmy', elements: companies },
    { scope: 'Pracownicy', elements: employees },
    { scope: 'Produkty', elements: products },
    { scope: 'Obiekty turystyczne', elements: tourismObjects },
    { scope: 'Menu', elements: menuItems },
  ];

  return (
    <div ref={dropdown}>
      <Dropdown isOpen={Boolean(showSuggestions && foundSuggestions)} toggle={toggle}>
        <div className="search-wrapper active">
          <div className="input-holder">
            <input
              data-t1="searchBox"
              onFocus={() => setShowSuggestions(true)}
              value={inputText}
              type="text"
              className="search-input"
              placeholder="Wpisz czego szukasz..."
              onChange={(e) => setInputText(e.target.value)}
            />
            <button type="button" className="search-icon">
              {isSearching ? <LoaderAnim color="#545cd8" type="ball-clip-rotate" active /> : <span />}
            </button>
          </div>
        </div>
        <DropdownToggle color="primary" tag="span" />
        <DropdownMenu className="dropdown-menu-hover-primary" data-t1="searchBoxResult">
          {suggestions.map(({ scope, elements }) => (elements && elements.length ? (
            <Fragment key={scope}>
              <DropdownItem header>{scope}</DropdownItem>
              {elements.map(({ to, label }, itemKey) => (
                <Link
                  to={to}
                  /* eslint-disable-next-line react/no-array-index-key */
                  key={itemKey}
                  className="text-decoration-none"
                  onClick={() => setShowSuggestions(false)}
                >
                  <DropdownItem>
                    {label}
                  </DropdownItem>
                </Link>
              ))}
            </Fragment>
          ) : null))}
        </DropdownMenu>
      </Dropdown>
    </div>
  );
}
