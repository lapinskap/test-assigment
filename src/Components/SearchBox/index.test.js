import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { mount } from 'enzyme';
import SearchBox, { findData } from './index';
import '../../tests/setupTests';
import TestAppContext from '../../tests/TestAppContext';
import { findCompanies } from './filters';
import { wait } from '../../tests/helpers';

describe('findData()', () => {
  it('should work correctly', async () => {
    const inputText = 'Fake Company';
    const setCompanies = jest.fn((result) => {
      expect(result.length).toBe(0);
    });
    await findData(inputText, setCompanies, findCompanies);
    expect(setCompanies).toBeCalledTimes(1);
    // Do not call function when empty input
    await findData('', setCompanies, findCompanies);
    expect(setCompanies).toBeCalledTimes(1);
  });
});

describe('<SearchBox />', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  global.document.createRange = () => ({
    setStart: () => {},
    setEnd: () => {},
    commonAncestorContainer: {
      nodeName: 'BODY',
      ownerDocument: document,
    },
  });

  it('renders correctly with default state', () => {
    const inputText = 'test input';
    const tree = renderer
      .create(<SearchBox inputText={inputText} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should Dropdown act correctly', async () => {
    const wrapper = mount(<SearchBox />);
    act(() => {
      const dropdown = wrapper.find('Dropdown').at(0);
      dropdown.simulate('click');
      wrapper.update();
      // should be false
      expect(wrapper.find('Dropdown').prop('isOpen')).toBe(false);
    });
    // should be true
    expect(wrapper.find('DropdownToggle').prop('aria-haspopup')).toBe(true);
  });

  it('should input setInputText method work', async () => {
    const wrapper = mount(
      <TestAppContext rbs router><SearchBox /></TestAppContext>,
    );
    const input = wrapper.find('input').at(0);
    act(() => {
      input.simulate('change', { target: { value: 'Lista' } });
    });
    wrapper.update();
    expect(wrapper.find('input').prop('value')).toBe('Lista');
    await wait(550); // wait to debounce
    wrapper.update();
    expect(wrapper.find('Dropdown').prop('isOpen')).toBe(true);

    act(() => {
      wrapper.find('Link').at(0).simulate('click');
    });
    wrapper.update();
    expect(wrapper.find('Dropdown').prop('isOpen')).toBe(false);
  });

  it('should input setShowSuggestions method work', async () => {
    const wrapper = mount(<SearchBox />);
    const input = wrapper.find('input').at(0);
    act(() => {
      input.simulate('focus');
      wrapper.update();
      expect(input).toBeTruthy();
    });
  });

  it('should dropdown item setShowSuggestions method work', async () => {
    const wrapper = mount(<SearchBox />);
    const dropdownItem = wrapper.find('Dropdown').at(0);
    act(() => {
      dropdownItem.simulate('click');
      wrapper.update();
      expect(dropdownItem.prop('isOpen')).toBe(false);
    });
    expect(dropdownItem.prop('disabled')).toBe(undefined);
  });
});
