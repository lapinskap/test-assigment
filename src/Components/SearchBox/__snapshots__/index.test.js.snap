// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBox /> renders correctly with default state 1`] = `
<div>
  <div
    className="dropdown"
    onKeyDown={[Function]}
  >
    <div
      className="search-wrapper active"
    >
      <div
        className="input-holder"
      >
        <input
          className="search-input"
          data-t1="searchBox"
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Wpisz czego szukasz..."
          type="text"
          value=""
        />
        <button
          className="search-icon"
          type="button"
        >
          <span />
        </button>
      </div>
    </div>
    <span
      aria-expanded={false}
      aria-haspopup={true}
      className=""
      onClick={[Function]}
    >
      <span
        className="sr-only"
      >
        Toggle Dropdown
      </span>
    </span>
    <div
      aria-hidden={true}
      className="dropdown-menu-hover-primary dropdown-menu"
      data-t1="searchBoxResult"
      role="menu"
      tabIndex="-1"
    />
  </div>
</div>
`;
